[Keymap] Add prog_qgmlwb keymap

davidkristoffersen opened the pull request:
## Description

Add the prog_qgmlwb keymap to the planck keyboard.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

drashna commented:
Please review the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and verify that you have followed the checklist for the type of PR this is.

Specifically, image files should be removed, and rehosted on Imgur or the like, if needed.  And any C/H files need to have gpl2+ compatible license headers.

davidkristoffersen commented:
I am now hosting the images on imgur, and have added spdx style gpl3 license headers to all *.c and *.h files. I do believe my contribution now follows the pr checklist.

davidkristoffersen commented:
Added changes to remove tracking of build folder and restructure/renaming of files

davidkristoffersen commented:
@drashna It has been a month since you requested the review, so I wondered if there is any progress on the review?

keyboard-magpie commented:
There is a heck of a lot going on here. The flash script seems unnecessary and certainly isn't something that's used in other mainstream keymaps.

Couldn't a lot of this be handled with userspace files?

davidkristoffersen commented:
I have restructured most of the code:
- Moved all utility like files into the userspace `users/davidkristoffersen`
- Added simple `README.md` files to each directory
- Added define guards to the userspace code
- Simplifying how multiple languages is handled in the keymap
- Removed printout and JSON files
- Moved all shell scripts into `scripts`

drashna commented on keyboards/planck/keymaps/prog_qgmlwb/scripts/config.sh:
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
```
Honestly, I'm not really a fan of having these scripts present, as a lot of this is outrigh handled by `qmk compile` and `qmk flash`, and the internal make processes.

drashna commented:
> Couldn't a lot of this be handled with userspace files?

Userspace is aimed more at sharing code between keyboards, especially when they're not similar/identical. 

Layouts is for sharing code and keymaps between boards that share the same layout config (eg "ortho_4x12").

drashna commented:
> I know most other keymaps don't use a flash script, I created it out of convenience for my own sake, and a bit since I'm used to creating such scripts for most of my applications and such. I prefer to keep it, but I understand if it must be removed.

Just a heads up, `qk_make` was added recently, and outputs something like:
```sh
qmk compile -kb handwired/tractyl_manuform/5x6_right/f411 -km drashna
qmk flash -kb handwired/tractyl_manuform/5x6_right/f411 -km drashna
```

However, it does require an extra define to enable.  Eg: 
```c
#define ENABLE_COMPILE_KEYCODE
```

Specifically, it uses the keyboard's name ("handwired/tractyl_manuform/5x6_right/f411" in the example), and the keymap name ("drashna" in the example), and automatically generates the command.   So this is "correct" no matter what keyboard you use. 

This may be close to what you want.

davidkristoffersen commented:
Changes for the next pr review:
- Renamed the `readme.md` files
- Removed the shell scripts and added usage of `QK_MAKE`(A great command actually!)
- Restructured some of the macro logic, such as using the `get_special_shifted_code` and `get_norwegian_code` as utility functions.
- Fixed some of the define guard logic, such as adding a `NO_SPECIAL_SHIFT` define
- I use both Norwegian and US keyboard layouts on my OS, that's why I'm making a layout that enables me to switch between them. I use the `get_special_shifted_code` logic to change certain shifted keys for US and Norwegian. I use the `get_norwegian_code` logic in order to avoid having to create a duplicate base, lower, and raise layer only for Norwegian, as most of the keys are the same.
